src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/AzureExportDestinationClient.cs

```diff
@@ -0,0 +1,104 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export.ExportDestinationClient
+{
+    public class AzureExportDestinationClient : IExportDestinationClient
+    {
+        private CloudBlobClient _blobClient = null;
+        private CloudBlobContainer _blobContainer = null;
+
+        private Dictionary<Uri, CloudBlockBlobWrapper> _uriToBlobMapping = new Dictionary<Uri, CloudBlockBlobWrapper>();
+        private Dictionary<(Uri FileUri, uint PartId), Stream> _streamMappings = new Dictionary<(Uri FileUri, uint PartId), Stream>();
+
+        public string DestinationType => "azure-block-blob";
+
+        public async Task ConnectAsync(string connectionSettings, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(connectionSettings, nameof(connectionSettings));
+
+            CloudStorageAccount cloudAccount = null;
+            if (!CloudStorageAccount.TryParse(connectionSettings, out cloudAccount))
+            {
+                throw new CantConnectToDestinationException();
+            }
+
+            _blobClient = cloudAccount.CreateCloudBlobClient();
+
+            // We will need to accept a container name/reference instead of using root container.
+            _blobContainer = _blobClient.GetRootContainerReference();
+            await _blobContainer.CreateIfNotExistsAsync();
+        }
+
+        public Task<Uri> CreateFileAsync(string fileName, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(fileName, nameof(fileName));
+            CheckIfClientIsConnected();
+
+            CloudBlockBlob blockBlob = _blobContainer.GetBlockBlobReference(fileName);
+            _uriToBlobMapping.Add(blockBlob.Uri, new CloudBlockBlobWrapper(blockBlob));
+
+            return Task.FromResult(blockBlob.Uri);
+        }
+
+        public async Task WriteFilePartAsync(Uri fileUri, uint partId, byte[] bytes, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(fileUri, nameof(fileUri));
+            EnsureArg.IsNotNull(bytes, nameof(bytes));
+            CheckIfClientIsConnected();
+
+            var key = (fileUri, partId);
```

Comment:
Now that I think about it, the key just need to be fileUri since commit will commit all of the parts and remove the mapping.

Reply:
I am going to leave it like this for now since we need the partId to generate the blockId. I will refactor it later.